### PR TITLE
WIP: Add axis=-1 as an argument to keras.backend.softmax

### PR DESCRIPTION
**Pull request by srjoglekar246**

### CODE REVIEW

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -1551,7 +1551,9 @@ def relu(x, alpha=0., max_value=None):
     return x
 
 
-def softmax(x):
+def softmax(x, axis=-1):
+    if axis != -1:
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: In general we can use the default implementation if `axis == 1 or axis == x.ndim - 1`

File: keras/backend/theano_backend.py [hunk 2]
```diff
@@ -1551,7 +1551,9 @@ def relu(x, alpha=0., max_value=None):
...
+        return T.exp(x) / T.exp(x).sum(axis=axis, keepdims=True)
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Could you try benchmarking this against the use of `transpose + softmax + transpose`? It may be that transposing is faster (but impossible to know in advance).

Comment (srjoglekar246) [reply keras/backend/theano_backend.py]: I am a little rusty with tensor shape transforms :-). Do you mean something like `T.nnet.softmax(x.transpose(0,1,3,2)).transpose(0,1,3,2)` ?

Comment (fchollet) [reply keras/backend/theano_backend.py]: The exact transposition depends on the `axis` value and `x.ndim`. But we'd only need to benchmark it for one configuration. For `axis=2` and `x.ndim=4` your code looks right to me.

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -1551,9 +1551,13 @@ def relu(x, alpha=0., max_value=None):
     return x
 
 
-def softmax(x):
-    return T.nnet.softmax(x)
-
+def softmax(x, axis=-1):
```
Comment (srjoglekar246) [inline keras/backend/theano_backend.py]: @fchollet This is what I could think of. Using transpose causes a "theano expects 1D/2D array, input is 3D" error for 3 dimensions.
This method gives 77ms mean time. The earlier one gave 60ms.

Comment (fchollet) [reply keras/backend/theano_backend.py]: Oh, you didn't have to actually implement the general case! Benchmarking ndim=2 and axis=0 would have been fine (and wouldn't require collapsing the array into 2D).

If the "manual" softmax computation is faster than the reshape/transpose, then let's go with that.

Thanks!

Comment (fchollet) [reply keras/backend/theano_backend.py]: Does CNTK support a softmax axis? If not, we can also go with manual softmax in CNTK.

Comment (srjoglekar246) [reply keras/backend/theano_backend.py]: I think the manual one is faster for some reason (atleast in the general n-dimensional case that I benchmarked anyway), so reverted to that.
CNTK does support axis for softmax: https://www.cntk.ai/pythondocs/cntk.ops.html?highlight=softmax#cntk.ops.softmax
